src/main/drivers/flash_w25n01g.c

```diff
@@ -0,0 +1,756 @@
+
+    READBYTES_DPRINTF(("readBytes: address 0x%x length %d\r\n", address, length));
+
+    uint32_t targetPage = W25N01G_LINEAR_TO_PAGE(address);
+
+    if (currentPage != targetPage) {
+        READBYTES_DPRINTF(("readBytes: PAGE_DATA_READ page 0x%x\r\n", targetPage));
+
+        cmd[0] = W25N01G_INSTRUCTION_PAGE_DATA_READ;
+        cmd[1] = 0;
+        cmd[2] = targetPage >> 8;
+        cmd[3] = targetPage;
+
+        if (!w25n01g_waitForReady(fdevice, W25N01G_TIMEOUT_PAGE_READ_MS)) {
+            return 0;
+        }
+
+        currentPage = UINT32_MAX;
+
+        ENABLE(fdevice->busdev);
+        spiTransfer(fdevice->busdev->busdev_u.spi.instance, cmd, NULL, 4);
+        DISABLE(fdevice->busdev);
+
+        if (!w25n01g_waitForReady(fdevice, W25N01G_TIMEOUT_PAGE_READ_MS)) {
+            return 0;
+        }
+
+        currentPage = targetPage;
+    }
+
+    uint16_t column = W25N01G_LINEAR_TO_COLUMN(address);
+    uint16_t transferLength;
+
+    if (length > W25N01G_PAGE_SIZE - column) {
+        transferLength = W25N01G_PAGE_SIZE - column;
+    } else {
+        transferLength = length;
+    }
+
+    cmd[0] = W25N01G_INSTRUCTION_READ_DATA;
+    cmd[1] = column >> 8;
+    cmd[2] = column;
+    cmd[3] = 0;
+
+    READBYTES_DPRINTF(("readBytes: READ_DATA column 0x%x transferLength 0x%x\r\n", column, transferLength));
+
+    ENABLE(fdevice->busdev);
+    spiTransfer(fdevice->busdev->busdev_u.spi.instance, cmd, NULL, 4);
+    spiTransfer(fdevice->busdev->busdev_u.spi.instance, NULL, buffer, length);
+    DISABLE(fdevice->busdev);
+
+    // XXX Don't need this?
+    if (!w25n01g_waitForReady(fdevice, W25N01G_TIMEOUT_PAGE_READ_MS)) {
+        return 0;
+    }
+
+    // Check ECC
+
+    uint8_t statReg = w25n01g_readRegister(fdevice->busdev, W25N01G_STAT_REG);
+    uint8_t eccCode = W25N01G_STATUS_FLAG_ECC(statReg);
+
+    switch (eccCode) {
+    case 0: // Successful read, no ECC correction
+        break;
+    case 1: // Successful read with ECC correction
+    case 2: // Uncorrectable ECC in a single page
+    case 3: // Uncorrectable ECC in multiple pages
+        w25n01g_addError(address, eccCode);
+        w25n01g_deviceReset(fdevice->busdev);
+        break;
+    }
+
+    READBYTES_DPRINTF(("readBytes: transfered 0x%x bytes\r\n", transferLength));
+
+    return transferLength;
+}
+
+int w25n01g_readExtensionBytes(flashDevice_t *fdevice, uint32_t address, uint8_t *buffer, int length)
+{
+    uint8_t cmd[4];
+
+    cmd[0] = W25N01G_INSTRUCTION_PAGE_DATA_READ;
+    cmd[1] = 0;
+    cmd[2] = W25N01G_LINEAR_TO_PAGE(address) >> 8;
+    cmd[3] = W25N01G_LINEAR_TO_PAGE(address);
+
+    ENABLE(fdevice->busdev);
+    spiTransfer(fdevice->busdev->busdev_u.spi.instance, cmd, NULL, 4);
+    DISABLE(fdevice->busdev);
+
+    if (!w25n01g_waitForReady(fdevice, W25N01G_TIMEOUT_PAGE_READ_MS)) {
+        return 0;
+    }
+
+    cmd[0] = W25N01G_INSTRUCTION_READ_DATA;
+    cmd[1] = 0;
+    cmd[2] = (2048 >> 8) & 0xff;
+    cmd[3] = 2048 & 0xff;
```

Comment:
This looks wrong.

From the data sheet, READ (W25N01G_INSTRUCTION_READ_DATA) = 03h, CA15-8, CA7-0, DUMMY.
But the above code has the dummy byte before the column address bytes.

Reply:
the `w25n01g_readBytes` method does this, correctly:

```
        uint8_t cmd[4];
        cmd[0] = W25N01G_INSTRUCTION_READ_DATA;
        cmd[1] = (column >> 8) & 0xff;
        cmd[2] = (column >> 0) & 0xff;
        cmd[3] = 0;
```